mlton/backend/ssa2-to-rssa.fun

```diff
@@ -1642,38 +1643,255 @@ fun convert (program as S.Program.T {functions, globals, main, ...},
+                     fun copy var' =
+                        (setGlobalStatic (var, globalStatic var');
+                         pushKeep st)
+                     fun keep () =
+                        pushKeep st
+                     fun word w =
+                        (setGlobalStatic (var, SOME (Word w));
+                         pushKeep st)
+                     fun real r =
+                        (setGlobalStatic (var, SOME (Real r));
+                         pushKeep st)
+                     fun 'a static (mk: S.Type.t * location * 'a ->
+                                    Var.t PackedRepresentation.staticOrElem, isEmpty, a: 'a) =
+                        let
+                           val location = getLocation (ty, isEmpty)
+                           val s = mk (ty, location, a)
+                           val rty = toRtype ty
+                        in
+                           case rty of
+                               SOME rty =>
+                                 (case (s, location) of
+                                      (* address of heap static not a valid elem *)
+                                      (Static s, Heap) =>
+                                       pushStatic (var, ty, rty, s)
+                                    | (Static s, _) =>
+                                       (setGlobalStatic (var, SOME (Address var));
+                                        pushStatic (var, ty, rty, s))
+                                    | (Elem e, _) =>
+                                       (setGlobalStatic (var, SOME e);
+                                        pushKeep st))
+                             | NONE => keep ()
+                        end
+                  in
+                     case exp of
+                        S.Exp.Const c =>
+                           (case c of
+                               Const.WordVector wxv => static (makeWordXVector, false, wxv)
+                             | Const.Word w => word w
+                             | Const.Real r => real r
+                             | _ => keep ())
+                      | S.Exp.Inject {variant, ...} =>
+                           copy variant
+                      | S.Exp.Object {args, con} =>
+
+                           let
+                              val location = getLocation (ty, false)
+                           in
+                              if Vector.exists (args, fn v =>
+                                        case globalStatic v of
+                                             NONE => true
+                                           | SOME (Real _) => RealX.disableCF ()
+                                           | _ => false)
+                              (* For objects, there's no reason to statically initalize,
+                               * since they're so small and irregular *)
+                              orelse location = Static.Heap
```

Comment:
Yes, it's just (hopefully)an optimization, since the code to initialize objects is quite small, making them into statics is likely to make the code size larger. Try switching the -static-alloc-objects flag on or off. In some cases even _with_ static allocation the code is larger than the old init code. Without it is just extra overhead. So it's a tradeoff which increases compile time but decreases code size.

Reply:
It is a little counter intuitive that static initialization leads to large code size that dynamic initialization.  But, I suppose that it is often due to the fact that initial field values are usually small constants (e.g., 0 or 1), which can be encoded in the instruction stream with an 8-bit immediates, rather than in the static data as a full 32- or 64-bit immediate.  Even an initial field value that is an objptr to another statically allocated object might end up in the instruction stream as a `%rip`-relative address (i.e., `&static_N + off`), which uses a 32-bit offset, while in the static data it would be a full 64-bit pointer (plus an entry in the relocation table, so that the dynamic linker can fixup the references in the data segment after the program is loaded).

I think this is why some systems actually create a little bytecode language for building the initial heap of literal objects.  For example, see SML/NJ's [new-literals.md](https://smlnj-gforge.cs.uchicago.edu/scm/viewvc.php/*checkout*/dev-notes/new-literals.md?revision=HEAD&root=smlnj) document.

Reply:
Resolved by 4a49d4d and deb6081.